openedx/openedx-platform
# WIP LX-74 Converts CAPA XModule to an XBlock

Converts CAPA XModule to a built-in XBlock, as was done for the Discussion XModule in https://github.com/edx/edx-platform/pull/12582.

**JIRA tickets**:  LX-74, [OSPR-2658](https://openedx.atlassian.net/browse/OSPR-2658)

**Dependencies**:

* https://github.com/edx/xblock-utils/pull/52
* https://github.com/jazkarta/edx-jsme/pull/7

**Screenshots**:

TBD -- goal is to keep GUI unchanged as much as possible.

**Sandbox URL**:

* LMS: https://pr19071.sandbox.opencraft.hosting/
* Studio: https://studio-pr19071.sandbox.opencraft.hosting/

(db0f8c8fcb3e9cf7bc59a0fb1e2eb8ba7dca9d27)

**Note** <strike>We're experiencing a bug with authentication for our sandboxes based on the latest master, and so to use the sandbox, you may need to work around this issue by logging into the Django Admin instead using the username (not email):</strike>  Bug fixed, workaround no longer required.

**Testing instructions**:

Before switching to this branch in your devstack:
1. Ensure that you have CAPA problems created in an existing course before switching to this branch.
   (Studio editing is not yet working.)  The edX Demo courses has most of them already.
1. In the LMS, choose one or more CAPA problems in an existing course.
1. Submit answers to these problems.
   We'll use this to ensure that the original submissions are not lost. 

Switch to this branch:
1. Run `paver update_assets` to update the python dependencies and rebuild the static assets.
1. Load the CAPA problems from above into the LMS.
    Check that your previous answers are saved.
1. Submit new answers.
   Ensure that these submissions are saved.
1. Check the remaining CAPA problem behaviours (Reset, Show Hint, ...)

**Author notes and concerns**:

1. This is a work-in-progress feature branch, and issues are being addressed.
   * Numerous `TODO`s in [openedx.core.lib.xblock_builtin.xblock_capa.xblock_capa.CapaXBlock](https://github.com/open-craft/edx-platform/blob/feature/capa_xblock/openedx/core/lib/xblock_builtin/xblock_capa/xblock_capa/__init__.py)
   * 6d88448 adds the `capa.scss` to the asset pipeline, which because it is a dependency of the [Conditional XModule](https://github.com/edx/edx-platform/blob/13a54b7046e11c9c26ca327ad564f06f1725a685/common/lib/xmodule/xmodule/conditional_module.py#L127), it remains in `common/static`.
      However, a number of changes had to be made to `capa.scss` to get it to build in the asset pipeline which may not be theme-compatible.  These need to be addressed.
   * Tests need to be fixed.
   * Need to handle LaTeX source editor.
   * Need to support editing and displaying complex capa problems, e.g. Chemical Equations and Molecule editors.
1. Have implemented a way to ensure the CapaXBlock OLX remains in a separate `problem` dir, but the code is pulled from several places, so could use a cleanup.
1. To continue to build JS/CSS by asset pipeline, and because some files are shared with other xmodules, these files were moved to the `common/static/common` dir.  Is it ok that these files are not consolidated under `openedx/core/lib/xblocks_builtin/xblock_capa`?
1. Translations will need to be regenerated at some point to reflect the moved files.  When/how should this be done?

**Reviewers**
- [ ] edX reviewer[s]: @ormsbee
